Add support for Amazon SES API

mabumusa1 opened the pull request:
<!--
Any PR related to Mautic 2 issues is not relavant anymore, please consider upgrading your code to the Mautic 3 series (staging/3.0 branch).
-->
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | staging
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Closes #1740, Resolves #7402, Mautic 3.x port of #6977

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-5-work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the staging branch.
-->

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do.
-->
#### Description:

<!--
If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. 

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

Commit:
Modify Test to include SES API

Commit:
Add Tests to SES API

	new file:   app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php
	modified:   composer.json
	modified:   composer.lock

Commit:
modified:   app/bundles/EmailBundle/Model/TransportType.php

wise-gorilla committed:
Return Path fix

wise-gorilla committed:
CS Fix

Commit:
Add automated tests

mabumusa1 commented:
@kuzmany may you please help me with this? I don't know why the coverage test fails

kuzmany commented:
@mabumusa1 this is duplicate of https://github.com/mautic/mautic/pull/6977 
Now community need decide to what PR is more complex and close to merge to core @dennisameling

mabumusa1 commented:
> @mabumusa1 this is duplicate of #6977
> Now community need decide to what PR is more complex and close to merge to core @dennisameling

@kuzmany I noticed that PR #6977 need a little work to get it done, so I talked to @Wise-Gorilla who managed to modify the original code written in #6977 to work the branch 3.x. so we created this new PR where we got @Wise-Gorilla code and added automated tests to it.

This is the same code @6977 but with changes to make it work with 3.x and some automated tests.

If there is a way we can work on the original PR, please let me know. it is better to have all the code in once place instead of having two PRs

kuzmany commented:
@mabumusa1 Thank you for clarification. Then make sense let this PR to tests and we can close https://github.com/mautic/mautic/pull/6977

KennedyTedesco commented:
Guys, this feature is a game-changer for AWS Users. Thanks for all effort put here. :heart:

kuzmany commented:
@KennedyTedesco  @edudemy Do you use this PR on production server? Write please more details, how many emails, what kind of sending services (campaign, broadcast emails) and if you didn't notice. Your reviews can speed up merging it to the core

heroic commented:
@kuzmany I am using the old PR 6977 in production. We're sending about 200k emails a day. Campaigns and segment emails.

mabumusa1 commented:
@heroic please write a review since you are using it, it will help us push this feature into production

chescos commented:
Thanks for working on this guys!

iurisilvio commented on app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php:
```diff
@@ -0,0 +1,861 @@
+        }
+
+        $BccAddresses = [];
+        if (count($messageArray['recipients']['cc']) > 0) {
```
Typo? cc -> bcc

wise-gorilla replied:
Thanks for the review, will fix that

iurisilvio commented on app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php:
```diff
@@ -0,0 +1,861 @@
+     */
+    public function getMaxBatchLimit()
+    {
+        return 50;
```
I'm not sure why, but using #6977 I have to set it as 49 to not exceed AWS limit.. Probably it is documented somewhere in AWS docs about what exactly counts as a recipient.

heroic replied:
I did not have to change this in #6977

wise-gorilla replied:
@heroic the original value was 10 on #6977, how was your sending speed?

heroic replied:
@Wise-Gorilla The send rate seems okay, coz our SES send limits are high(500/s, and 1M per day). But our emails aren't really time sensitive so doesn't matter as much to us.

wise-gorilla replied:
Actually what i meant is when that value was set to 10 on #6977 then how many emails per second you was able to send approximately?

mabumusa1 replied:
I think the limit of 50 is set because in SES, you can send to 50 recipients for each batch sent, mentioned here https://docs.aws.amazon.com/ses/latest/DeveloperGuide/quotas.html check the part that says `Maximum number of recipients per message`

RCheesley replied:
@mabumusa1 @Wise-Gorilla again, please resolve the conversation if it's dealt with if not, a status update would be helpful.

dennisameling replied:
I'm leaving the limit at 50 as per the official Amazon docs. Since Amazon counts anyone in the To, CC and BCC fields against this limit, it might cause issues if Mautic e.g. looks at only the "To" field. But let's merge this for now. If anyone runs into problems with this setting, please feel free to submit a bug report. Mautic will throw an exception if this happens so it's easy to troubleshoot then.

iurisilvio commented on app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php:
```diff
@@ -0,0 +1,861 @@
+        if (count($this->templateCache)) {
+            $this->logger->debug('Deleting SES templates that were created in this session');
+            foreach ($this->templateCache as $templateName) {
+                $this->deleteSesTemplate($templateName);
```
https://github.com/mautic/mautic/pull/6977#issuecomment-519614524

Didn't tested this PR, but running #6977 this delete crashed the integration. I think AWS takes some time to process the emails and we remove the template before they finish processing.

I'm running without deleting templates. It is not perfect, but better than unexpected crashes.

```
[2019-08-08 17:02:55] mautic.ERROR: Exception sending email template: TemplateDoesNotExist, Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist [] []
[2019-08-08 17:02:55] mautic.ERROR: [MAIL ERROR] Error executing "SendBulkTemplatedEmail" on "https://email.us-east-1.amazonaws.com"; AWS HTTP error: Client error: `POST https://email.us-east-1.amazonaws.com` resulted in a `400 Bad Request` response: <ErrorResponse xmlns="http://ses.amazonaws.com/doc/2010-12-01/">   <Error>     <Type>Sender</Type>     <Code>TemplateDoe (truncated...)  TemplateDoesNotExist (client): Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist - <ErrorResponse xmlns="http://ses.amazonaws.com/doc/2010-12-01/">   <Error>     <Type>Sender</Type>     <Code>TemplateDoesNotExist</Code>     <Message>Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist</Message>   </Error>   <RequestId>fa6e5393-3d5b-4089-a3f3-7774213227a8</RequestId> </ErrorResponse>  (send); REDACTED_EMAILS_WERE_HERE {"exception":"[object] (Exception(code: 0): Error executing \"SendBulkTemplatedEmail\" on \"https://email.us-east-1.amazonaws.com\"; AWS HTTP error: Client error: `POST https://email.us-east-1.amazonaws.com` resulted in a `400 Bad Request` response:\n<ErrorResponse xmlns=\"http://ses.amazonaws.com/doc/2010-12-01/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>TemplateDoe (truncated...)\n TemplateDoesNotExist (client): Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist - <ErrorResponse xmlns=\"http://ses.amazonaws.com/doc/2010-12-01/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>TemplateDoesNotExist</Code>\n    <Message>Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist</Message>\n  </Error>\n  <RequestId>fa6e5393-3d5b-4089-a3f3-7774213227a8</RequestId>\n</ErrorResponse>\n at /var/www/html/app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php:457)"} []
``

heroic replied:
With #6977 we did not have to remove these lines. Works fine for me. We had to add a pid to fix the error mentioned. From what the Author and I gathered, it seemed 2 processes were sending emails in parallel, and a race condition occurred which deleted the template before the second process sent emails, but after the second process found the templates to already exist.

iurisilvio replied:
Great! Glad you already investigated and handled the issue. I'll try this solution in my production code.

mabumusa1 replied:
@heroic may you please add the lines that solved this issue?

wise-gorilla replied:
Guess you are refering to adding the getmypid() to the template name like what i did in line 502 to make the template name more unique, right?

heroic replied:
@Wise-Gorilla yes.

heroic replied:
https://github.com/mautic/mautic/pull/9005/files#diff-327931ffb6a20b14b8da54a78284787cR502

RCheesley replied:
@Wise-Gorilla @mabumusa1 was this resolved? Please resolve the conversation if so.

dennisameling replied:
Looks like this was resolved. I'm resolving the conversation so we can go ahead and merge 👍

iurisilvio commented:
I already commented in the past PR, but this feature is critical to AWS SES. It sends a lot faster.

I'm sending 80k emails with rate limit 100/s in less than one hour. I don't know exactly the time spent, AWS don't have stats per minute).

My Mautic is in Brazil and AWS SES in Virginia. With this setup, we sent 5k emails/hour through SMTP, impossible to use with 80k emails 3 times per week, including some time sensitive offers.

iurisilvio commented on app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php:
```diff
@@ -0,0 +1,861 @@
+     *
+     * @return array
+     */
+    public function processCallbackRequest(Request $request)
```
All these webhook processing below here are almost identical to `AmazonTransport`.

I have little experience with Mautic/PHP code, but looks like it could be extracted to a common file to avoid repetition.

wise-gorilla replied:
Thanks Luri for that suggestion, I did that and refactored the callback methods and moved it to a separate class.

wise-gorilla committed:
Updated SES regions

Updated SES regions to include all avilable 14 regions for both SMTP and API and changed the translation to use more user frindly region names.

wise-gorilla committed:
Fix bcc typo

wise-gorilla committed:
Add replyTo and returnPath to RawMessage

wise-gorilla committed:
Changed the ConfigurationSetName header name

Changed the ConfigurationSetName to X-SES-CONFIGURATION-SET as mentioned here
https://docs.aws.amazon.com/ses/latest/DeveloperGuide/using-configuration-sets-in-email.html

wise-gorilla committed:
CS Fix

wise-gorilla committed:
Add SES/SNS bounces to email graph #7893

Add SES/SNS bounces to email graph, Now the  API transport uses the same callback code used in SMTP transport

dennisameling commented:
As asked by @Wise-Gorilla [on Slack](https://mautic.slack.com/archives/C02HV76BW/p1595837937439700):

> i'm working on Add support to SES API #9005 PR, The AmazonApiTransport is using the same callback methods that used in AmazonTransport so i'm trying to move the Callback methods (processCallbackRequest, processJsonPayload, processBounce, processUnsubscription, getSnsPayload) to a separate class to avoid dublicated code and tried to mimic the way used in SendGridTransport however i get this error, Can anyone help me solving this issue please?
```
[2020-07-26 22:51:43] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalThrowableError: "Type error: Argument 4 passed to Mautic\EmailBundle\Swiftmailer\Transport\AmazonApiTransport::__construct() must be an instance of Mautic\EmailBundle\Swiftmailer\Amazon\AmazonCallback, instance of Mautic\EmailBundle\Model\TransportCallback given, called in /www/wwwroot/m3.wise-gorilla.ga/var/cache/prod/ContainerEhllbsd/getMautic_Transport_AmazonApiService.php on line 8" at /www/wwwroot/m3.wise-gorilla.ga/app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php line 121 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Type error: Argument 4 passed to Mautic\\EmailBundle\\Swiftmailer\\Transport\\AmazonApiTransport::__construct() must be an instance of Mautic\\EmailBundle\\Swiftmailer\\Amazon\\AmazonCallback, instance of Mautic\\EmailBundle\\Model\\TransportCallback given, called in /www/wwwroot/m3.wise-gorilla.ga/var/cache/prod/ContainerEhllbsd/getMautic_Transport_AmazonApiService.php on line 8 at /www/wwwroot/m3.wise-gorilla.ga/app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php:121)"} []
```

You are getting this error because you are sending the `TransportCallback` to the function instead of `AmazonCallback`, see [config.php line 403](https://github.com/mautic/mautic/blob/4123e0322909e1293853052b57a42f981c191d01/app/bundles/EmailBundle/Config/config.php#L403):

```PHP
            'mautic.transport.amazon_api' => [
                'class'        => 'Mautic\EmailBundle\Swiftmailer\Transport\AmazonApiTransport',
                'serviceAlias' => 'swiftmailer.mailer.transport.%s',
                'arguments'    => [
                    'mautic.http.connector',
                    'monolog.logger.mautic',
                    'translator',
                    'mautic.email.model.transport_callback', // ===== THIS LINE NEEDS TO BE CHANGED
                ],
                'methodCalls' => [
                    'setUsername' => ['%mautic.mailer_user%'],
                    'setPassword' => ['%mautic.mailer_password%'],
                    'setRegion'   => ['%mautic.mailer_amazon_api_region%'],
                ],
            ],
```

You'll need to register your `AmazonCallback` in the `config.php` just like `mautic.email.model.transport_callback`, then you can add it as an argument to the `AmazonApiTransport` constructor.

You also asked the following on Slack:

> do you recommend to do this refactoring?

If code is duplicated in the SMTP and API versions, then yes please refactor. You can also look at some other integrations (like Sendgrid) which also have their own Callbacks.

wise-gorilla committed:
Amazon callback refactoring

Refactored the callback for AmazonApiTransport and AmazonTransport and moved the dublicated methods in a separate callback class.

Jammmmm commented:
I like some others thought "Amazon SES" meant using the API. Only now after attempting to send to 200,000+ subscribers did we find out it meant SMTP via Amazon instead.

Realistically, does this PR look like it's on track for inclusion into 3.1.0? (16th August, 2020)

Commit:
Fixes to the tests

Commit:
Fixes to the tests

wise-gorilla committed:
Refactoring fix

wise-gorilla commented:
> I like some others thought "Amazon SES" meant using the API. Only now after attempting to send to 200,000+ subscribers did we find out it meant SMTP via Amazon instead.
> 
> Realistically, does this PR look like it's on track for inclusion into 3.1.0? (16th August, 2020)



> I like some others thought "Amazon SES" meant using the API. Only now after attempting to send to 200,000+ subscribers did we find out it meant SMTP via Amazon instead.
> 
> Realistically, does this PR look like it's on track for inclusion into 3.1.0? (16th August, 2020)

@Jammmmm If you helped us by reviewing and testing this PR then it will find it's way to Mautic 3.1.

wise-gorilla committed:
AmazonCallback test refactoring

wise-gorilla committed:
Automated test fix

wise-gorilla committed:
Automated test fix

wise-gorilla committed:
Automated test fix

wise-gorilla committed:
CS Fix

wise-gorilla committed:
Removed unused methods and use keywords

wise-gorilla committed:
Show Amazon SES regions option during installation

wise-gorilla committed:
CS Fix

wise-gorilla committed:
CS Fix

wise-gorilla commented:
@RCheesley @dennisameling 
I addressed the issue #7160  and PR #8735  and i'm about to merge them to add the solusion to this PR. After merging the users will be able to have an "Other"option on the region box, for Amazon SMTP choosing the Other option will show the regular SMTP host field and on Amazon API choosing other will show a new text field to add the region.

I also made the SMTP port to be always visible for Amazon SMTP for all regions including "Other" option.

All that worked find but i had a small cosmotic issue because of that.

This is how the fields look like now
![image](https://user-images.githubusercontent.com/65309837/89120042-0ec13900-d4b3-11ea-95d9-4b42ad3d8e75.png)

However if you used any other regions it will look like this, notice the port field position.. i understand why it look like this and the field position can be edited from _config_emailconfig_widget.html.php but i don't know what i can do to solve that ugly positioning.

![image](https://user-images.githubusercontent.com/65309837/89120052-2e586180-d4b3-11ea-8d9c-1e0295d4c67a.png)

and even if i moved the port field to the region row it will look nice on Amazon SMTP transport like this
![image](https://user-images.githubusercontent.com/65309837/89120243-184ba080-d4b5-11ea-936b-77269b9f0a70.png)

However it will look bad on "Other SMTP Transport" like this
![image](https://user-images.githubusercontent.com/65309837/89120259-35806f00-d4b5-11ea-8b1b-6711f06929e6.png)

I think one of the solusion to solve that is to create a new text field for amazon port and host which will give me more flexability in term of positioning instead of reusing the existing Host and Port field but i don't think it's a good idea.

Hope somone can help me so we can close those #7160 and  #8735

wise-gorilla committed:
Configration fields enhancments

Added Other region option to the dropdown menu and allows to set a custom port for SMTP transport

kuzmany commented on app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php:
```diff
@@ -0,0 +1,703 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Should be 2020

kuzmany commented on app/bundles/EmailBundle/Swiftmailer/Amazon/AmazonCallback.php:
```diff
@@ -0,0 +1,227 @@
+<?php
+
+/*
```
Copyright on new files should have actual year

kuzmany commented on app/bundles/EmailBundle/Config/config.php:
```diff
@@ -383,14 +383,29 @@
+                ],
+            ],
+            'mautic.transport.amazon_api' => [
+                'class'        => 'Mautic\EmailBundle\Swiftmailer\Transport\AmazonApiTransport',
```
Should be \Mautic\EmailBundle\Swiftmailer\Transport\AmazonApiTransport::class

wise-gorilla replied:
Thanks so much for your review, regards this edit i'm a bit confused since original amazon transport uses
`'Mautic\EmailBundle\Swiftmailer\Transport\AmazonTransport'`
and all other transports use the same format as well so i'm wondering if ther is a specific reason for your suggestion.

kuzmany replied:
It's just minor. Full class reference is better to read by IDE software and refactor it.

wise-gorilla replied:
Thanks Kuzmany, I just pushed your edits, Please let me know if you suggest any other edits.

kuzmany reviewed (dismissed):
Notices just minor things

wise-gorilla committed:
CS Fix

wise-gorilla committed:
Full class reference

wise-gorilla committed:
Revert "Full class reference"

This reverts commit daa75e243e44279a022aac66a173a9b105d99c7a.

wise-gorilla committed:
CS Fix

wise-gorilla committed:
CS Fix

dennisameling commented:
@Wise-Gorilla did you know you can run `bin/php-cs-fixer fix --config=.php_cs --verbose --using-cache=no` to automatically fix CS issues? 😊 

Normally this should run automatically with a Git hook just before you commit your changes, but I guess that Git hook isn't working for you 🙈

wise-gorilla commented:
> @Wise-Gorilla did you know you can run `bin/php-cs-fixer fix --config=.php_cs --verbose --using-cache=no` to automatically fix CS issues? 😊
> 
> Normally this should run automatically with a Git hook just before you commit your changes, but I guess that Git hook isn't working for you 🙈

Thanks @dennisameling for the advice :) i'm still very new to Github and i'm not familiar yet to Git hooks, I will need to learn about it :)

wise-gorilla committed:
Consolidate smtp and api region dropdown

wise-gorilla committed:
CS Fix

wise-gorilla committed:
Auto test fix

wise-gorilla committed:
Auto test fix

wise-gorilla committed:
Fix conflict

RCheesley commented:
@Wise-Gorilla @mabumusa1 looks like we have a conflict that needs to be resolved - please can you take a look?

RCheesley reviewed (commented):
Hi folks, can we get a status update on this PR? I know a *lot* of work has gone into it and a *lot* of people stand to benefit.

Scrolling up I saw several unresolved conversations from previous discussions - please mark them as resolved if they are fixed?

It would be extremely helpful if we could get the conflict resolved, get the code coverage report and then make sure we have two good tests with the PR in this finalised state, so that we can get it into 3.1. Let's make it happen! 🚀

dennisameling committed:
Fix test and conflicts

dennisameling reviewed (approved):
This is great news for Amazon SES users. Thanks @Wise-Gorilla @mabumusa1 @idealaunchtech and everyone else involved for pushing this forward!

I'd like to urgently ask @Wise-Gorilla @mabumusa1 to add automated tests to make the code easier to understand for other developers and to increase stability of the feature/product. I'll go ahead and merge this now for the 3.1 RC release tomorrow so folks can already test functionality; if you could add tests in the course of the coming week we can get those merged into the 3.1 final release 👍 

Tracking this in https://github.com/mautic/mautic/issues/9092
